File: zabbix_add_hosts_java/code.java

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.InetAddress;
import java.net.URL;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ZabbixHostCreator {

    private static final String ZABBIX_URL = System.getenv("ZABBIX_URL");
    private static final String ZABBIX_USER = System.getenv("ZABBIX_USER");
    private static final String ZABBIX_PASSWORD = System.getenv("ZABBIX_PASSWORD");
    private static final String DNS_FILE = System.getenv("DNS_FILE");
    private static final String HOST_GROUP_NAME = System.getenv("HOST_GROUP");
    private static final ObjectMapper mapper = new ObjectMapper();


    public static void main(String[] args) {
        try {
            String token = getZabbixToken();
            String hostGroupId = getHostGroupId(token);

            try (BufferedReader reader = new BufferedReader(new FileReader(DNS_FILE))) {
                String dns;
                while ((dns = reader.readLine()) != null) {
                     dns = dns.trim();
                     if (dns.isEmpty()){
                         continue;
                     }

                    try {
                         String ip = getIpByDns(dns);
                         createHost(token, dns, ip, hostGroupId);
                     } catch (IOException e) {
                          System.err.println("Error during processing DNS " + dns + ": " + e.getMessage());
                     }
                }
            }


            System.out.println("Hosts creation finished.");
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
        }
    }

     private static String getIpByDns(String dns) throws IOException{
        try{
            InetAddress address = InetAddress.getByName(dns);
            return address.getHostAddress();
        } catch (Exception e){
            throw new IOException("Error getting ip address from dns: " + dns + e.getMessage());
        }
     }

    private static String getZabbixToken() throws IOException {
        Map<String, Object> authRequest = new HashMap<>();
        authRequest.put("jsonrpc", "2.0");
        authRequest.put("method", "user.login");
        Map<String, String> params = new HashMap<>();
        params.put("user", ZABBIX_USER);
        params.put("password", ZABBIX_PASSWORD);
        authRequest.put("params", params);
        authRequest.put("id", 1);

        String response = sendRequest(ZABBIX_URL, authRequest);
        Map<?, ?> jsonResponse = mapper.readValue(response, Map.class);

        if (jsonResponse.containsKey("error")) {
             Map<?, ?> error = (Map<?, ?>) jsonResponse.get("error");
            throw new IOException("Zabbix error: "+ error.get("message").toString() + ", code:"+ error.get("code").toString() + ", data:"+error.get("data").toString());
        }

        return jsonResponse.get("result").toString();
    }

    private static String getHostGroupId(String token) throws IOException {
        Map<String, Object> hostGroupRequest = new HashMap<>();
        hostGroupRequest.put("jsonrpc", "2.0");
        hostGroupRequest.put("method", "hostgroup.get");
        Map<String, Object> params = new HashMap<>();
        Map<String, String> filter = new HashMap<>();
        filter.put("name", HOST_GROUP_NAME);
        params.put("filter", filter);
        hostGroupRequest.put("params", params);
        hostGroupRequest.put("id", 1);

        String response = sendRequestWithAuth(ZABBIX_URL, hostGroupRequest, token);
        Map<?, ?> jsonResponse = mapper.readValue(response, Map.class);

        if (jsonResponse.containsKey("error")) {
             Map<?, ?> error = (Map<?, ?>) jsonResponse.get("error");
            throw new IOException("Zabbix error: "+ error.get("message").toString() + ", code:"+ error.get("code").toString() + ", data:"+error.get("data").toString());
        }


        List<?> result = (List<?>) jsonResponse.get("result");

        if (result != null && !result.isEmpty()) {
            Map<?,?> group = (Map<?,?>) result.get(0);
            if(group.containsKey("groupid")){
              return  group.get("groupid").toString();
            }

        }
        throw new IOException("Group with name '" + HOST_GROUP_NAME + "' not found.");
    }


     private static void createHost(String token, String dns, String ip, String hostGroupId) throws IOException {
        Map<String, Object> hostRequest = new HashMap<>();
        hostRequest.put("jsonrpc", "2.0");
        hostRequest.put("method", "host.create");


        Map<String, Object> params = new HashMap<>();
        params.put("host", dns);
        params.put("name", dns);
        List<Map<String, Object>> interfaces = Arrays.asList(createInterface(ip));
        params.put("interfaces", interfaces);

        List<Map<String, Object>> groups = Arrays.asList(createHostGroup(hostGroupId));
        params.put("groups", groups);

        hostRequest.put("params", params);
        hostRequest.put("id", 1);



        String response = sendRequestWithAuth(ZABBIX_URL, hostRequest, token);
        Map<?, ?> jsonResponse = mapper.readValue(response, Map.class);

          if (jsonResponse.containsKey("error")) {
             Map<?, ?> error = (Map<?, ?>) jsonResponse.get("error");
            throw new IOException("Zabbix error: "+ error.get("message").toString() + ", code:"+ error.get("code").toString() + ", data:"+error.get("data").toString());
        }


        if (jsonResponse.containsKey("result") &&  ((Map<?, ?>)jsonResponse.get("result")).containsKey("hostids")){
          System.out.println("Host '" + dns + "' created with ID: "+ ((List<?>)((Map<?, ?>)jsonResponse.get("result")).get("hostids")).get(0).toString());
        }else{
           throw  new IOException("Error create host " + dns + " error: "+ jsonResponse.get("result").toString());
        }

    }

    private static Map<String, Object> createInterface(String ip) {
        Map<String, Object> iface = new HashMap<>();
        iface.put("type", 1);
        iface.put("main", 1);
        iface.put("useip", 1);
        iface.put("ip", ip);
        iface.put("dns", "");
        iface.put("port", "10050");
        return iface;
    }


    private static Map<String, Object> createHostGroup(String groupId) {
        Map<String, Object> group = new HashMap<>();
        group.put("groupids", Arrays.asList(groupId));
        return group;
    }


    private static String sendRequest(String url, Map<String, Object> request) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setDoOutput(true);


        String jsonInputString = mapper.writeValueAsString(request);

        try (OutputStream os = connection.getOutputStream()) {
             byte[] input = jsonInputString.getBytes("utf-8");
             os.write(input, 0, input.length);
        }


        StringBuilder response = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(connection.getInputStream(), "utf-8"))) {
            String responseLine;
            while ((responseLine = br.readLine()) != null) {
                response.append(responseLine.trim());
            }
        }
        connection.disconnect();

        return response.toString();
    }

     private static String sendRequestWithAuth(String url, Map<String, Object> request, String token) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setRequestProperty("Authorization", "Bearer "+token);
        connection.setDoOutput(true);


         String jsonInputString = mapper.writeValueAsString(request);

        try (OutputStream os = connection.getOutputStream()) {
            byte[] input = jsonInputString.getBytes("utf-8");
            os.write(input, 0, input.length);
        }


        StringBuilder response = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(connection.getInputStream(), "utf-8"))) {
            String responseLine;
            while ((responseLine = br.readLine()) != null) {
                response.append(responseLine.trim());
            }
        }

         connection.disconnect();

        return response.toString();
    }
}

File: code.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/joho/godotenv"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

// Zabbix API Request Structs
type ZabbixAuthRequest struct {
	Jsonrpc string `json:"jsonrpc"`
	Method  string `json:"method"`
	Params  struct {
		User     string `json:"user"`
		Password string `json:"password"`
	} `json:"params"`
	ID int `json:"id"`
}

type ZabbixRequest struct {
	Jsonrpc string      `json:"jsonrpc"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params"`
	ID      int         `json:"id"`
}

// Zabbix API Response Structs
type ZabbixResponse struct {
	Jsonrpc string      `json:"jsonrpc"`
	Result  interface{} `json:"result,omitempty"`
	Error   struct {
		Code    int    `json:"code"`
		Message string `json:"message"`
		Data    string `json:"data"`
	} `json:"error,omitempty"`
	ID int `json:"id"`
}
type ZabbixHost struct {
	Hostid string `json:"hostid"`
	Host   string `json:"host"`
	Name   string `json:"name"`
}

// Utility function to load environment variables from .env file
func loadEnv() {
	err := godotenv.Load(".env")
	if err != nil {
		log.Fatalf("Error loading .env file")
	}
}

func getZabbixToken(url, user, password string) (string, error) {
	authRequest := ZabbixAuthRequest{
		Jsonrpc: "2.0",
		Method:  "user.login",
		Params: struct {
			User     string `json:"user"`
			Password string `json:"password"`
		}{
			User:     user,
			Password: password,
		},
		ID: 1,
	}
	jsonAuthRequest, err := json.Marshal(authRequest)
	if err != nil {
		return "", fmt.Errorf("error marshalling json request : %w", err)
	}
	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequest("POST", url, bytes.NewReader(jsonAuthRequest))
	if err != nil {
		return "", fmt.Errorf("error creating request : %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("error performing request : %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("error reading response body: %w", err)
	}
	var zabbixResponse ZabbixResponse
	if err := json.Unmarshal(body, &zabbixResponse); err != nil {
		return "", fmt.Errorf("error unmarshaling response body : %w body:%s", err, string(body))
	}
	if zabbixResponse.Error.Message != "" {
		return "", fmt.Errorf("zabbix error: %s, code:%d, data:%s", zabbixResponse.Error.Message, zabbixResponse.Error.Code, zabbixResponse.Error.Data)
	}

	token, ok := zabbixResponse.Result.(string)
	if !ok {
		return "", fmt.Errorf("unexpected response format: %v", zabbixResponse.Result)
	}
	return token, nil
}

func getHostGroupId(token, url, groupName string) (string, error) {
	params := map[string]interface{}{
		"filter": map[string]interface{}{
			"name": groupName,
		},
	}
	request := ZabbixRequest{
		Jsonrpc: "2.0",
		Method:  "hostgroup.get",
		Params:  params,
		ID:      1,
	}

	jsonData, err := json.Marshal(request)
	if err != nil {
		return "", fmt.Errorf("error marshalling hostgroup request: %w", err)
	}

	client := http.Client{
		Timeout: 10 * time.Second,
	}
	req, err := http.NewRequest("POST", url, bytes.NewReader(jsonData))
	if err != nil {
		return "", fmt.Errorf("error creating hostgroup request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("error sending hostgroup request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("error reading hostgroup response: %w", err)
	}
	var zabbixResponse ZabbixResponse
	if err := json.Unmarshal(body, &zabbixResponse); err != nil {
		return "", fmt.Errorf("error unmarshalling hostgroup response: %w, body:%s", err, string(body))
	}
	if zabbixResponse.Error.Message != "" {
		return "", fmt.Errorf("zabbix error: %s, code:%d, data:%s", zabbixResponse.Error.Message, zabbixResponse.Error.Code, zabbixResponse.Error.Data)
	}

	if result, ok := zabbixResponse.Result.([]interface{}); ok && len(result) > 0 {
		if group, ok := result[0].(map[string]interface{}); ok {
			if id, ok := group["groupid"].(string); ok {
				return id, nil
			}
		}
	}

	return "", fmt.Errorf("group with name '%s' not found", groupName)
}

func getHostsByGroupId(token, url, groupId string) ([]ZabbixHost, error) {
	params := map[string]interface{}{
		"output":   []string{"hostid", "host", "name"},
		"groupids": []string{groupId},
	}
	request := ZabbixRequest{
		Jsonrpc: "2.0",
		Method:  "host.get",
		Params:  params,
		ID:      1,
	}

	jsonData, err := json.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("error marshalling host get request: %w", err)
	}

	client := http.Client{
		Timeout: 10 * time.Second,
	}
	req, err := http.NewRequest("POST", url, bytes.NewReader(jsonData))
	if err != nil {
		return nil, fmt.Errorf("error creating host get request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error sending host get request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading host get response: %w", err)
	}
	var zabbixResponse ZabbixResponse
	if err := json.Unmarshal(body, &zabbixResponse); err != nil {
		return nil, fmt.Errorf("error unmarshalling host get response: %w, body:%s", err, string(body))
	}
	if zabbixResponse.Error.Message != "" {
		return nil, fmt.Errorf("zabbix error: %s, code:%d, data:%s", zabbixResponse.Error.Message, zabbixResponse.Error.Code, zabbixResponse.Error.Data)
	}

	var hosts []ZabbixHost
	if result, ok := zabbixResponse.Result.([]interface{}); ok {
		for _, hostData := range result {
			if host, ok := hostData.(map[string]interface{}); ok {
				hostItem := ZabbixHost{
					Hostid: host["hostid"].(string),
					Host:   host["host"].(string),
					Name:   host["name"].(string),
				}
				hosts = append(hosts, hostItem)
			}
		}
	}

	return hosts, nil
}
func updateHost(token, url, hostId, newName string) (bool, error) {
	params := map[string]interface{}{
		"hostid": hostId,
		"name":   newName,
	}

	request := ZabbixRequest{
		Jsonrpc: "2.0",
		Method:  "host.update",
		Params:  params,
		ID:      1,
	}

	jsonData, err := json.Marshal(request)
	if err != nil {
		return false, fmt.Errorf("error marshalling host update request: %w", err)
	}
	client := http.Client{
		Timeout: 10 * time.Second,
	}
	req, err := http.NewRequest("POST", url, bytes.NewReader(jsonData))
	if err != nil {
		return false, fmt.Errorf("error creating host update request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := client.Do(req)
	if err != nil {
		return false, fmt.Errorf("error sending host update request: %w", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return false, fmt.Errorf("error reading host update response: %w", err)
	}

	var zabbixResponse ZabbixResponse
	if err := json.Unmarshal(body, &zabbixResponse); err != nil {
		return false, fmt.Errorf("error unmarshalling host update response: %w, body:%s", err, string(body))
	}
	if zabbixResponse.Error.Message != "" {
		return false, fmt.Errorf("zabbix error: %s, code:%d, data:%s", zabbixResponse.Error.Message, zabbixResponse.Error.Code, zabbixResponse.Error.Data)
	}

	if result, ok := zabbixResponse.Result.(map[string]interface{}); ok {
		if _, ok := result["hostids"].([]interface{}); ok {
			return true, nil
		}
	}
	return false, fmt.Errorf("error update host: %v", zabbixResponse.Result)
}

func main() {
	loadEnv()
	zabbixURL := os.Getenv("ZABBIX_URL")
	zabbixUser := os.Getenv("ZABBIX_USER")
	zabbixPassword := os.Getenv("ZABBIX_PASSWORD")
	hostGroupName := os.Getenv("HOST_GROUP")

	token, err := getZabbixToken(zabbixURL, zabbixUser, zabbixPassword)
	if err != nil {
		log.Fatalf("Error getting Zabbix token: %v", err)
	}
	log.Println("Successfully logged in to Zabbix")

	hostGroupId, err := getHostGroupId(token, zabbixURL, hostGroupName)
	if err != nil {
		log.Fatalf("Error getting host group ID: %v", err)
	}

	hosts, err := getHostsByGroupId(token, zabbixURL, hostGroupId)
	if err != nil {
		log.Fatalf("Error getting hosts by group id: %v", err)
	}

	log.Printf("Found %d hosts in group %s", len(hosts), hostGroupName)

	for _, host := range hosts {
		newName := host.Name + ".isb"
		ok, err := updateHost(token, zabbixURL, host.Hostid, newName)
		if err != nil {
			log.Printf("Error update host %s with id %s: %v", host.Name, host.Hostid, err)
			continue
		}
		if ok {
			log.Printf("Host '%s' updated with new name: %s", host.Host, newName)
		}
	}

	log.Println("Hosts updated finished.")
}
